### PR TITLE
[GH-1105] Update workloads in background periodically

### PR DESCRIPTION
**Pull request by ehigham**
## Purpose
https://broadinstitute.atlassian.net/browse/GH-1105

Implement the `update-workloads!` function called by the "workload-manager" (i.e. the background loop that keeps the workload database up-to-date).
- Only update "live" workloads (i.e. those that are not `:finished`)

Remove updating workloads from the `api/v1/workload` endpoint handler
- This is now done periodically.
- Still guarantees eventual consistency like the lazy implementation before.

In the test tools, get workload updates from workflow-launcher itself instead of talking to terra/cromwell directly.

## Questions
- Not sure how to best handle exceptions in the update loop. I'm expecting some db conflicts but not much else. If we can't connect to the database, we probably don't want to bring the server down. If there's a data issue, maybe we do?

### CODE REVIEW

File: api/src/wfl/service/postgres.clj [hunk 1]
```diff
@@ -42,6 +42,15 @@
...
     (throw (ex-info (format "Table %s does not exist" table) {:cause "no-such-table"}))))
 
+(defn ^:private cromwell-status
+  "`status` of the workflow with `uuid` in `cromwell`."
+  [cromwell uuid]
+  (-> (str/join "/" [cromwell "api" "workflows" "v1" uuid "status"])
+      (http/get {:headers (once/get-auth-header)})
+      :body
+      util/parse-json
+      :status))
```
Comment (tbl3rd) [inline api/src/wfl/service/postgres.clj:52]: Consolidate this cromwell-status with the one in workloads.clj?

File: api/src/wfl/api/workloads.clj [hunk 2]
```diff
@@ -64,13 +64,13 @@
...
   (let [do-load   (partial load-workload-impl tx)]
-    (map do-load (jdbc/query tx ["SELECT * FROM workload WHERE project = ?" project]))))
+    (mapv do-load (jdbc/query tx ["SELECT * FROM workload WHERE project = ?" project]))))
 
 (defn load-workloads
   "Use transaction TX to load all known `workloads`"
   [tx]
   (let [do-load (partial load-workload-impl tx)]
-    (map do-load (jdbc/query tx ["SELECT * FROM workload"]))))
+    (mapv do-load (jdbc/query tx ["SELECT * FROM workload"]))))
```
Comment (tbl3rd) [inline api/src/wfl/api/workloads.clj:73]: Why the `mapv`s?

Comment (tbl3rd) [reply api/src/wfl/api/workloads.clj:73]: Ah, to force query completion within the transaction?

File: api/test/wfl/tools/workloads.clj [hunk 3]
```diff
@@ -206,13 +199,9 @@
...
     (wfl.api.workloads/load-workload-for-id tx id)))
 
-;; `doall` is required here for testing otherwise the moment test uses
-;; the result and tries to force realizing, the db tx is already closed
-;;
 (defn load-workloads-with-project [project]
   (jdbc/with-db-transaction [tx (fixtures/testing-db-config)]
-    (doall
-     (wfl.api.workloads/load-workloads-with-project tx project))))
+    (wfl.api.workloads/load-workloads-with-project tx project)))
```
Comment (tbl3rd) [inline api/test/wfl/tools/workloads.clj:204]: The `mapv`s explained.

File: api/src/wfl/server.clj [hunk 4]
```diff
@@ -98,9 +116,10 @@
       (isCancelled [_] false)
       (isDone [_] (.isStopped server)))))
```
Comment (tbl3rd) [inline api/src/wfl/server.clj:117]: The start-workload-manager docstring is in the "wrong" place. Intentional?

**Review by tbl3rd — APPROVED** [review]
Better and simpler than without the background updating. Nice.

File: api/src/wfl/service/postgres.clj (hunk 1 above)
Comment (ehigham) [reply api/src/wfl/service/postgres.clj:52]: Good spot! That one isn't used so I'll just remove it!

File: api/src/wfl/api/workloads.clj (hunk 2 above)
Comment (ehigham) [reply api/src/wfl/api/workloads.clj:73]: Exactly - I could have added `doall`s everywhere but I think eager evaluation is used more commonly.

File: api/src/wfl/server.clj (hunk 4 above)
Comment (ehigham) [reply api/src/wfl/server.clj:117]: Thanks!

**Pull request merged**